mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java

```diff
@@ -56,10 +61,28 @@ public void initialize(@Nullable Configuration configuration, Properties serDePr
     } else if (serDeProperties.get(InputFormatConfig.TABLE_SCHEMA) != null) {
       tableSchema = SchemaParser.fromJson((String) serDeProperties.get(InputFormatConfig.TABLE_SCHEMA));
     } else {
-      try {
-        tableSchema = Catalogs.loadTable(configuration, serDeProperties).schema();
-      } catch (NoSuchTableException nte) {
-        throw new SerDeException("Please provide an existing table or a valid schema", nte);
+      if (Catalogs.hiveCatalog(configuration)) {
+        tableSchema = hiveSchema(serDeProperties);
+        if (tableSchema == null) {
+          throw new SerDeException("Please provide a valid schema");
+        } else {
+          LOG.info("Using hive schema {}", SchemaParser.toJson(tableSchema));
+        }
+      } else {
+        try {
+          // always prefer the original table schema if there is one
+          tableSchema = Catalogs.loadTable(configuration, serDeProperties).schema();
+          LOG.info("Using schema from existing table {}", SchemaParser.toJson(tableSchema));
+        } catch (Exception e) {
+          // If we can not load the table try the provided hive schema
+          tableSchema = hiveSchema(serDeProperties);
+          if (tableSchema == null) {
+            throw new SerDeException("Please provide an existing table or a valid schema", e);
+          } else {
+            LOG.info("Using schema from column specification {} since table load is failed",
+                SchemaParser.toJson(tableSchema), e);
```

Comment:
Can we extract this out in a common function to reuse above?

Reply:
Done